### ISSUE TITLE
RegisterComponents() for a list of Module

### ISSUE THREAD

**Issue by QingtaoLi1** [dotnet/TorchSharp#330]
I would like to use a Dictionary to store submodules in my model, but RegisterComponents() only take effect for Module and Tensor fields. Shall I use something like TorchSharp.torch.nn.Sequential to wrap the submodules in order to make them registered? And to my understanding there may be another solution -- adding detection of List and Dictionary in RegisterComponents(). (I'm not sure whether this will break any existing code logic.)

**Comment by NiklasGustafsson** [comment]
Hi @QingtaoLi1,

A small example that illustrates what you are trying to do would be really helpful.

**Comment by QingtaoLi1** [comment]
@NiklasGustafsson OK, I need to define a Dictionary of Modules in the model, but I find that `RegisterComponents()` will not register these modules because it only deals with fields of type Tensor and Module.
An example:
```
public class Model : torch.nn.CustomModule
{
    public Dictionary<string, torch.nn.Module> ClassificationHeads;
    public Model() : base("Model")
    {
        {Define modules and initialization;}
        RegisterComponents();
    }
}
```

**Comment by NiklasGustafsson** [comment]
@QingtaoLi1 Thanks for that clarification. Can you educate me on why you need the dictionary? Is there a dynamic number of submodules? Why does it not work to have a field for each submodule?

Thanks,

Niklas

**Comment by QingtaoLi1** [comment]
@NiklasGustafsson Thanks for your question. In the example we need different submodules for different NLP tasks. It is possible to define a field for each submodule, however if we want to add a brand new task, a new field should be added. Using a dictionary would be more elegant.
In another model class, we need a fixed number of layers (say 24). If we use direct fields, there will be 24 different layer fields, which also disables for-loop to operate these layers. In my opinion, it is natural to use a List to include them.

**Comment by NiklasGustafsson** [comment]
I'm not sure I understand why you wouldn't just have different models and share parameters, but I'll take your word for it.

I suggest making the `Module.state_dict()` function virtual, and then you can override it, so both save() and load() will use your implementation instead of the native code one. It does have to be a string->Tensor dictionary, though, not a list.

```CSharp
public Dictionary<string, Tensor> state_dict() { ... }
```

As far as the List scenario goes, that's what Sequential is for.

**Comment by NiklasGustafsson** [comment]
This is addressed in PR #355, which virtualizes state_dict() and load(BinaryReader). With that, a custom module may save and restore anything it desires to.
